streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -704,6 +707,10 @@ void runOnce() {
                     numIterations++;
                 }
             } while (true);
+
+            // we record the ratio out of the while loop so that the accumulated latency spans over
+            // multiple iterations with reasonably large max.num.records and hence is less vulnerable to outliers
+            taskManager.recordTaskProcessRatio(totalProcessLatency);
```

Comment:
Note we should only call this inside the `if (state == State.RUNNING)` condition otherwise we may get ConcurrentModification exception.

Reply:
Q: Could you elaborate on this?

Reply:
I do not know exactly what's happening here, my guess is that in PARTITION_REVOKED state the tasks might be modified.